### ISSUE TITLE
real_sense_grabber.cpp fails to compile under VS2015 (pcl_io)

### ISSUE THREAD

**Issue by nznobody** [PointCloudLibrary/pcl#1557]
Hello, first I'd like to apologize for anything I do wrong, this is my first time using GIT, let alone PCL. I read the guidelines and to my understanding this is OK.

I am attempting to compile the latest master under VS2015 (have done dependencies) and ran into this overloaded operator issue:

[First Erroneous Line](https://github.com/PointCloudLibrary/pcl/blame/master/io/src/real_sense_grabber.cpp#L267)

This is also present on line 282 and 449

VS2015 gives the following error:

```
C:\Code\pcl\io\src\real_sense_grabber.cpp(267): error C2678: binary '==': no operator found which takes a left-hand operand of type 'pcl::RealSenseGrabber::Mode' (or there is no acceptable conversion)
5>  C:\Code\pcl\common\include\pcl/PCLHeader.h(46): note: could be 'bool pcl::operator ==(const pcl::PCLHeader &,const pcl::PCLHeader &)' (compiling source file C:\Code\pcl\io\src\real_sense_grabber.cpp)
5>  C:\Code\pcl\io\src\real_sense_grabber.cpp(267): note: while trying to match the argument list '(pcl::RealSenseGrabber::Mode, pcl::RealSenseGrabber::Mode)'
```

I have temporarily fixed this for me by overloading the pcl::RealSenseGrabber::Mode == operator manually. I can then fix line 282 by casting them manually.

Does anyone have a better solution?

**Comment by taketwo** [comment]
Looks weird, comparison operator is defined at [line 106](https://github.com/PointCloudLibrary/pcl/blame/master/io/src/real_sense_grabber.cpp#L106). Is this the entire error message that you get? I would expect one more line with explanation of why exactly it failed to match the argument list...

**Comment by UnaNancyOwen** [comment]
I has confirmed that it reproduce in my environment.
The following is environment and build log of pcl_io.
* PCL master/HEAD
* Visual Studio 2015 Update1
* RealSense SDK 2016 R1
```
1>------ Build started: Project: pcl_io, Configuration: Release x64 ------
1>  real_sense_grabber.cpp
1>C:\pcl-master\common\include\pcl/point_traits.h : warning C4819: The file contains a character that cannot be represented in the current code page (932). Save the file in Unicode format to prevent data loss
1>C:\pcl-master\io\src\real_sense_grabber.cpp(267): error C2678: binary '==': no operator found which takes a left-hand operand of type 'pcl::RealSenseGrabber::Mode' (or there is no acceptable conversion)
1>  C:\pcl-master\common\include\pcl/PCLHeader.h(46): note: could be 'bool pcl::operator ==(const pcl::PCLHeader &,const pcl::PCLHeader &)'
1>  C:\pcl-master\io\src\real_sense_grabber.cpp(267): note: while trying to match the argument list '(pcl::RealSenseGrabber::Mode, pcl::RealSenseGrabber::Mode)'
1>C:\pcl-master\io\src\real_sense_grabber.cpp(282): error C2678: binary '==': no operator found which takes a left-hand operand of type 'const pcl::RealSenseGrabber::Mode' (or there is no acceptable conversion)
1>  C:\pcl-master\common\include\pcl/PCLHeader.h(46): note: could be 'bool pcl::operator ==(const pcl::PCLHeader &,const pcl::PCLHeader &)'
1>  C:\pcl-master\io\src\real_sense_grabber.cpp(282): note: while trying to match the argument list '(const pcl::RealSenseGrabber::Mode, pcl::RealSenseGrabber::Mode)'
1>C:\pcl-master\io\src\real_sense_grabber.cpp(449): error C2678: binary '==': no operator found which takes a left-hand operand of type 'pcl::RealSenseGrabber::Mode' (or there is no acceptable conversion)
1>  C:\pcl-master\common\include\pcl/PCLHeader.h(46): note: could be 'bool pcl::operator ==(const pcl::PCLHeader &,const pcl::PCLHeader &)'
1>  C:\pcl-master\io\src\real_sense_grabber.cpp(449): note: while trying to match the argument list '(pcl::RealSenseGrabber::Mode, pcl::RealSenseGrabber::Mode)'
========== Build: 0 succeeded, 1 failed, 3 up-to-date, 0 skipped ==========
```
```
Error	C2678	binary '==': no operator found which takes a left-hand operand of type 'pcl::RealSenseGrabber::Mode' (or there is no acceptable conversion)	pcl_io	C:\pcl-master\io\src\real_sense_grabber.cpp	267	
Error	C2678	binary '==': no operator found which takes a left-hand operand of type 'const pcl::RealSenseGrabber::Mode' (or there is no acceptable conversion)	pcl_io	C:\pcl-master\io\src\real_sense_grabber.cpp	282	
Error	C2678	binary '==': no operator found which takes a left-hand operand of type 'pcl::RealSenseGrabber::Mode' (or there is no acceptable conversion)	pcl_io	C:\pcl-master\io\src\real_sense_grabber.cpp	449	
```


I think that should define the operator== in Mode (<code>pcl::RealSenseGrabber::Mode::operator==</code>).
I has confirmed that build is successful by apply this fixes.

* [real_sense_grabber.h#L85-L112](https://github.com/PointCloudLibrary/pcl/blob/master/io/include/pcl/io/real_sense_grabber.h#L85-L112)
```cpp
struct PCL_EXPORTS Mode
{
  /* ... */
  bool
  operator== (const pcl::RealSenseGrabber::Mode& m) const;
};
```

* [real_sense_grabber.cpp#L105-L113](https://github.com/PointCloudLibrary/pcl/blob/master/io/src/real_sense_grabber.cpp#L105-L113)
```cpp
bool
pcl::RealSenseGrabber::Mode::operator== (const pcl::RealSenseGrabber::Mode& m) const
{
  return (this->fps == m.fps &&
          this->depth_width == m.depth_width &&
          this->depth_height == m.depth_height &&
          this->color_width == m.color_width &&
          this->color_height == m.color_height);
}
```

Do you have any good ideas?

**Comment by taketwo** [comment]
I don't understand why the original code fails to compile and how your proposed change is fundamentally different from it. That being said, if this fix for some reason does make the code compile, then let's apply it. Pull request is welcome.

**Comment by UnaNancyOwen** [comment]
I was send a pull request #1560 .
Please review it.

I has confirmed that build is successful by apply this fixes.
But, I can't operation test, because I not have the Intel RealSense camera.
If you have RealSense camera, Please check the operation.

Thanks.

**Comment by nznobody** [comment]
> BTW: 
Although not directly related to this issue, FindRSSDK.cmake always will find x64 library.
CMake should be find Win32 library when target platform is x86.
Now, It has change settings manually to Win32 when building MSVC x86.


Can confirm this too. Building PCL under VS 2015 has its quirks ;)

**Comment by UnaNancyOwen** [comment]
In addition, Fix to find RealSense SDK library corresponding to target platform.
Please review pull request.

**Comment by taketwo** [comment]
Fixed via #1560.
